// src/models/tokenProfiles/basicIdToken.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';

var jwtDecoder = require('../../controllers/messageTypes/jwt/jsonwebtoken/decode');
var jwtSigner = require('../../controllers/messageTypes/jwt/jsonwebtoken/sign');

/* Init token using standard claims */ 
function BasicIdToken(iss, sub, iat, jti){
    this.initData();    
    this.iss = iss;
    this.sub = sub;
    this.iat = iat;
    this.jti = jti
    this.validateRequiredFields();
};

/* Provided standard claims */
BasicIdToken.prototype.standard_claims = {};

/* Provided non standard claims */ 
BasicIdToken.prototype.non_standard_claims = {};

/* Expected standard claim values */
BasicIdToken.prototype.verification_claims = {};

/* Expected non standard verification claims that are known */
BasicIdToken.prototype.non_standard_verification_claims = {};

BasicIdToken.prototype.NoneAlgorithm = false;

/* Required standard claims */
BasicIdToken.prototype.options_to_payload = {
    'iss': 'iss',
    'sub': 'sub',
    'iat': 'iat',
    'jti': 'jti',
};
  
/* Other option values */ 
BasicIdToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

/* Known standard claims */
BasicIdToken.prototype.knownNonStandardClaims = {
    'aud': 'aud',
    'exp': 'exp',
    'nbf': 'nbf',
};

/* Required verification claims */
BasicIdToken.prototype.claims_to_verify = {
    'iss': 'iss',
    'sub': 'sub',
    'maxAge' : 'maxAge',
    'jti': 'jti',
};

BasicIdToken.prototype.initData = function(){
    BasicIdToken.prototype.non_standard_verification_claims = {};    
    BasicIdToken.prototype.NoneAlgorithm = false;
};

BasicIdToken.prototype.validateRequiredFields = function(){

    if (this.iss && this.sub && this.iat && this.jti){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

BasicIdToken.prototype.addNonStandardClaims = function(nonStandardClaims){
    BasicIdToken.prototype.non_standard_claims = nonStandardClaims;

    BasicIdToken.prototype.non_standard_verification_claims = {};
    Object.keys(nonStandardClaims).forEach(function (key) {
        if (BasicIdToken.prototype.knownNonStandardClaims[key]) {
            BasicIdToken.prototype.non_standard_verification_claims[key] = nonStandardClaims[key];
        }
    });  
};

BasicIdToken.prototype.getStandardClaims = function(){
    BasicIdToken.prototype.standard_claims = { "iss" : this.iss, "sub" : this.sub, "iat": this.iat, "jti": this.jti};
    return BasicIdToken.prototype.standard_claims;         
};

BasicIdToken.prototype.getNonStandardClaims = function(nonStandardClaims){
    return BasicIdToken.prototype.non_standard_claims;
}; 

BasicIdToken.prototype.getVerificationClaims = function(){
    return BasicIdToken.prototype.verification_claims;
}; 

BasicIdToken.prototype.getNonStandardVerificationClaims = function(){
    return BasicIdToken.prototype.non_standard_verification_claims;
}; 

/* User explicitly wants to set None Algorithm attribute */
BasicIdToken.prototype.setNoneAlgorithm = function(boolVal){
    BasicIdToken.prototype.NoneAlgorithm = boolVal;
};

BasicIdToken.prototype.getNoneAlgorithm = function(boolVal){
    return BasicIdToken.prototype.NoneAlgorithm;
};

/* Serialization of JWT type */
BasicIdToken.prototype.toJWT = function(secretOrPrivateKey, options, callback){
    return jwtSigner.sign(this, secretOrPrivateKey, options, callback);
};

/* Deserialization of JWT type */
BasicIdToken.prototype.fromJWT = function(signedJWT, secretOrPrivateKey, claimsToVerify, options, callback){

    this.validateRequiredVerificationClaims(claimsToVerify);
    this.validateRequiredNonStandardVerificationClaims(claimsToVerify);
    return jwtDecoder.decode(signedJWT,secretOrPrivateKey, this, options, callback);
};

/* Throws error if required verification claims are not present */ 
BasicIdToken.prototype.validateRequiredVerificationClaims = function(claimsToVerify)
{
    Object.keys(BasicIdToken.prototype.claims_to_verify).forEach(function (key) {
        if (!claimsToVerify[key]) {
            throw new Error('Missing required verification claim: ' + key);
        }
      });  
    BasicIdToken.prototype.verification_claims = claimsToVerify;
};

/* Throws error if required non standard verification claims are not present */ 
BasicIdToken.prototype.validateRequiredNonStandardVerificationClaims = function(claimsToVerify)
{
    if (BasicIdToken.prototype.non_standard_verification_claims['nbf'] || BasicIdToken.prototype.non_standard_verification_claims['exp']){
        this.nonStandardVerificationClaimsCheck('clockTolerance', claimsToVerify);
    }
    if (BasicIdToken.prototype.non_standard_verification_claims['aud']){
        this.nonStandardVerificationClaimsCheck('aud', claimsToVerify);
    }
};

BasicIdToken.prototype.nonStandardVerificationClaimsCheck = function(key, claimsToVerify){
    if (!claimsToVerify[key]) {
        throw new Error('Missing required verification claim: ' + key);
    }else{
        BasicIdToken.prototype.verification_claims[key] = claimsToVerify[key];
        if (key == "aud"){
            BasicIdToken.prototype.claims_to_verify['aud'] = 'aud';
        }
    }
}

BasicIdToken.prototype.toJSON = function(){
    // TODO
    var obj = Object.assign({}, this.getStandardClaims(), this.getNonStandardClaims());
    return JSON.stringify(obj);
};

BasicIdToken.prototype.fromJSON = function(jsonString){
    // TODO
    return JSON.parse(jsonString);
};

BasicIdToken.prototype.toUrlEncoded = function(){
    // TODO
    var obj = Object.assign({}, this.getStandardClaims(), this.getNonStandardClaims());
    var str = [];
    for(var p in obj)
        str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
    return str.join("&");
}

BasicIdToken.prototype.fromUrlEncoded = function(urlEncodedString){
    // TODO
    var obj = {}; 
    urlEncodedString.replace(/([^=&]+)=([^&]*)/g, function(m, key, value) {
        obj[decodeURIComponent(key)] = decodeURIComponent(value);
    }); 
    return obj;
};

module.exports = BasicIdToken;
