// src/models/tokenProfiles/extendedIdToken.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';

var GoogleIdToken = require('./googleIdToken');  
var jwtDecoder = require('../../controllers/messageTypes/jwt/jsonwebtoken/decode');
var jwtSigner = require('../../controllers/messageTypes/jwt/jsonwebtoken/decode');

/* Init token using standard claims */ 
function ExtendedIdToken(name, email, picture, iss, sub, iat){
    this.initData();
    this.name = name;
    this.email = email;
    this.picture = picture;
    this.iss = iss;
    this.sub = sub;
    this.iat = iat;
    this.validateRequiredFields();
};

ExtendedIdToken.prototype = Object.create(GoogleIdToken.prototype);
ExtendedIdToken.prototype.constructor = ExtendedIdToken;

/* Required standard claims */
ExtendedIdToken.prototype.options_to_payload = {
    'name': 'name',
    'email': 'email',
    'picture': 'picture',
    'iss': 'iss',
    'sub': 'sub',
    'iat': 'iat',
};
  
/* Other options values */
ExtendedIdToken.prototype.options_for_objects = [
    'expiresIn',
    'notBefore',
    'noTimestamp',
    'audience',
    'issuer',
    'subject',
    'jwtid',
];

/* Known non standard claims to be verified */
ExtendedIdToken.prototype.knownNonStandardClaims = {
    'aud': 'aud',
    'exp': 'exp',
    'nbf': 'nbf',
};

/* Required standard claims to be verified */
ExtendedIdToken.prototype.claims_to_verify = {
    'name': 'name',
    'email': 'email',
    'picture': 'picture',
    'iss': 'iss',
    'sub': 'sub',
    'maxAge' : 'maxAge',
};

ExtendedIdToken.prototype.validateRequiredFields = function(){
    if (this.name && this.email && this.picture && this.iss && this.sub && this.iat){
        console.log("Validated all standard fields")
    }else {
        throw new Error("You are missing a required parameter");
    }
};

ExtendedIdToken.prototype.getStandardClaims = function(){
    ExtendedIdToken.prototype.standard_claims = {"name": this.name, "email" : this.email, "picture": this.picture,  "iss" : this.iss, "sub" : this.sub, "iat": this.iat};
    return ExtendedIdToken.prototype.standard_claims;         
};

ExtendedIdToken.prototype.initData = function(){
    ExtendedIdToken.prototype.non_standard_verification_claims = {};    
    ExtendedIdToken.prototype.NoneAlgorithm = false;
};

ExtendedIdToken.prototype.addNonStandardClaims = function(nonStandardClaims){
    ExtendedIdToken.prototype.non_standard_claims = nonStandardClaims;

    ExtendedIdToken.prototype.non_standard_verification_claims = {};
    Object.keys(nonStandardClaims).forEach(function (key) {
        if (ExtendedIdToken.prototype.knownNonStandardClaims[key]) {
            ExtendedIdToken.prototype.non_standard_verification_claims[key] = nonStandardClaims[key];
        }
    });  
};
ExtendedIdToken.prototype.getNonStandardClaims = function(nonStandardClaims){
    return ExtendedIdToken.prototype.non_standard_claims;
};


ExtendedIdToken.prototype.getVerificationClaims = function(){
    return ExtendedIdToken.prototype.verification_claims;
}; 

ExtendedIdToken.prototype.getNonStandardVerificationClaims = function(){
    return ExtendedIdToken.prototype.non_standard_verification_claims;
};

/* Serialization of JWT type */ 
ExtendedIdToken.prototype.fromJWT = function(signedJWT, secretOrPrivateKey, claimsToVerify, options){
        this.validateRequiredVerificationClaims(claimsToVerify);
        this.validateRequiredNonStandardVerificationClaims(claimsToVerify);
        return jwtDecoder.decode(signedJWT,secretOrPrivateKey, this, options);
};

/* Check for required verification claims that need to be verified */
ExtendedIdToken.prototype.validateRequiredVerificationClaims = function(claimsToVerify)
{
    Object.keys(ExtendedIdToken.prototype.claims_to_verify).forEach(function (key) {
        if (!claimsToVerify[key]) {
            throw new Error('Missing required verification claim: ' + key);
        }
      });  
      ExtendedIdToken.prototype.verification_claims = claimsToVerify;
};

/* Check for required non standard verification claims that need to be verified */
ExtendedIdToken.prototype.validateRequiredNonStandardVerificationClaims = function(claimsToVerify)
{
    if (ExtendedIdToken.prototype.non_standard_verification_claims['exp']){
        this.nonStandardVerificationClaimsCheck('clockTolerance', claimsToVerify);
    }
    if (ExtendedIdToken.prototype.non_standard_verification_claims['aud']){
        this.nonStandardVerificationClaimsCheck('aud', claimsToVerify);
    }

};

ExtendedIdToken.prototype.nonStandardVerificationClaimsCheck = function(key, claimsToVerify){
    if (!claimsToVerify[key]) {
        throw new Error('Missing required verification claim: ' + key);
    }else{
        ExtendedIdToken.prototype.verification_claims[key] = claimsToVerify[key];
        if (key == "aud"){
            ExtendedIdToken.prototype.claims_to_verify['aud'] = 'aud';
        }
    }
}


module.exports = ExtendedIdToken;
