// get-pkg-repo/index.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';
var parseSlug = require('parse-github-repo-url');
var normalizeData = require('normalize-package-data');
var hostedGitInfo = require('hosted-git-info');
var url = require('url');
var typos = require('./typos');

var gitAt = /^git@/;

function gitAtToHttpsUrl(repoUrl) {
  repoUrl = repoUrl.replace(':', '/');
  repoUrl = repoUrl.replace(gitAt, 'https://');
  return repoUrl;
}

function parseRepoUrl(repoUrl) {
  var parsed = url.parse(repoUrl);
  if (!parsed.host) {
    var gitAt = /^git@/;
    if (gitAt.test(repoUrl)) {
      parsed = url.parse(gitAtToHttpsUrl(repoUrl));
    }
  }
  return parsed;
}

function getType(repoUrl) {
  if (repoUrl.indexOf('github') !== -1) {
    return 'github';
  }
  if (repoUrl.indexOf('gitlab') !== -1) {
    return 'gitlab';
  }
}

function unknownHostedInfo(repoUrl) {
  var parsed = parseRepoUrl(repoUrl);
  var protocol = parsed.protocol === 'https:' ? 'https:' : 'http:';
  var browseUrl = protocol + '//' + (parsed.host || '') + parsed.path.replace(/\.git$/, '').replace(/\/$/, '');

  var UnknownGitHost = function() {
    var slug = parseSlug(repoUrl);

    if (parsed.host) {
      this.domain = parsed.host;
    }

    this.user = slug[0];
    this.project = slug[1];

    this.type = getType(repoUrl);
  };

  UnknownGitHost.prototype.browse = function() {
    return browseUrl;
  };

  return new UnknownGitHost();
}

function getPkgRepo(pkgData, fixTypo) {
  try {
    pkgData = JSON.parse(pkgData);
  } catch (err) {}

  if (fixTypo && !pkgData.repository) {
    typos.forEach(function(val) {
      if (pkgData[val]) {
        pkgData.repository = pkgData[val];
        return false;
      }
    });
  }
  normalizeData(pkgData);

  var repo = pkgData.repository;
  if (!repo || !repo.url) {
    throw new Error('No "repository" field found in your package.json file.' +
      ' Please see https://docs.npmjs.com/files/package.json#repository for proper syntax.');
  }

  return hostedGitInfo.fromUrl(repo.url) || unknownHostedInfo(repo.url);
}

module.exports = getPkgRepo;
