// src/controllers/messageTypes/jwt/lib/jws/lib/sign-stream.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

/*global module*/
var base64url = require('base64url');

var DataStream = require('./data-stream');
var jwa = require('../../jwa');
var Stream = require('stream');
var toString = require('./tostring');
var util = require('util');

var conv = require('binstring');
var base32 = require('hi-base32');

function jwsSecuredInputBase(header, payload, encoding) {
  encoding = encoding || 'utf8';
  var encodedHeader = base64url(toString(header), 'binary');
  var encodedPayload = base64url(toString(payload), encoding);
  return util.format('%s.%s', encodedHeader, encodedPayload);
}

function jwsSecuredInputBase16(header, payload, encoding) {
  encoding = encoding || 'utf8';
  var uriEncoded = encodeURIComponent(JSON.stringify(header));
  var uriEncodedPayload = encodeURIComponent(JSON.stringify(payload));
  var encodedHeader = conv(uriEncoded, { in:'binary', out:'hex' });
  var encodedPayload = conv(uriEncodedPayload, { in:encoding, out:'hex' });
  return util.format('%s.%s', encodedHeader, encodedPayload);
}

function jwsSecuredInputBase32(header, payload, encoding) {
  encoding = encoding || 'utf8';
  var uriEncoded = encodeURIComponent(JSON.stringify(header));
  var uriEncodedPayload = encodeURIComponent(JSON.stringify(payload));
  var encodedHeader = base32.encode(new Buffer(uriEncoded, 'binary'));
  var encodedPayload = base32.encode(new Buffer(uriEncodedPayload, encoding));
  return util.format('%s.%s', encodedHeader, encodedPayload);
}

function jwsSign(opts) {
  var header = opts.header;
  var payload = opts.payload;
  var secretOrKey = opts.secret || opts.privateKey;
  var encoding = opts.encoding;
  var baseEncoding = opts.baseEncoding;
  var algo = jwa(header.alg);

  var securedInput = "";
  switch(baseEncoding) {
    case "base16":
        securedInput = jwsSecuredInputBase16(header, payload, encoding);
        break;
    case "base32":
        securedInput = jwsSecuredInputBase32(header, payload, encoding);
        break;
    default:
        securedInput = jwsSecuredInputBase(header, payload, encoding);
  }

  var signature = algo.sign(securedInput, secretOrKey, baseEncoding);
  return util.format('%s.%s', securedInput, signature);
}

function SignStream(opts) {
  var secret = opts.secret||opts.privateKey||opts.key;
  var secretStream = new DataStream(secret);
  this.readable = true;
  this.header = opts.header;
  this.encoding = opts.encoding;
  this.secret = this.privateKey = this.key = secretStream;
  this.payload = new DataStream(opts.payload);
  this.secret.once('close', function () {
    if (!this.payload.writable && this.readable)
      this.sign();
  }.bind(this));

  this.payload.once('close', function () {
    if (!this.secret.writable && this.readable)
      this.sign();
  }.bind(this));
}
util.inherits(SignStream, Stream);

SignStream.prototype.sign = function sign() {
  try {
    var signature = jwsSign({
      header: this.header,
      payload: this.payload.buffer,
      secret: this.secret.buffer,
      encoding: this.encoding, 
      baseEncoding: this.baseEncoding,
    });
    this.emit('done', signature);
    this.emit('data', signature);
    this.emit('end');
    this.readable = false;
    return signature;
  } catch (e) {
    this.readable = false;
    this.emit('error', e);
    this.emit('close');
  }
};

SignStream.sign = jwsSign;

module.exports = SignStream;
