// asyncawait/src/async/config.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

var assert = require('assert');
/** A config object holds the configuration options for a variant of the async function. */
var Config = (function () {
    /** Construct a new Config instance. */
    function Config(options) {
        /** Recognised values: 'none', 'promise', 'thunk', 'result'. */
        this.returnValue = Config.PROMISE;
        /** Indicates whether a callback function, if supplied, will be used to notify waiters of results. */
        this.acceptsCallback = false;
        /** Indicates whether the suspendable function has iterator semantics or normal semantics. */
        this.isIterable = false;
        /** Indicates whether top-level concurrency should be limited to a specified ceiling. */
        this.maxConcurrency = null;
        if (options) {
            this.returnValue = options.returnValue;
            this.acceptsCallback = options.acceptsCallback;
            this.isIterable = options.isIterable;
            this.maxConcurrency = options.maxConcurrency;
        }
    }
    /** Checks all configuration values and throw an error if anything is invalid. */
    Config.prototype.validate = function () {
        var knownRetVal = [Config.PROMISE, Config.THUNK, Config.RESULT, Config.NONE].indexOf(this.returnValue) !== -1;
        assert(knownRetVal, 'Unrecognised return value: ' + this.returnValue);
        var hasNotifier = this.returnValue !== Config.NONE || this.acceptsCallback;
        assert(hasNotifier, 'At least one notification method must be enabled.');
    };
    // Constants for use with returnValue and callbackArg
    Config.PROMISE = 'promise';
    Config.THUNK = 'thunk';
    Config.RESULT = 'result';
    Config.NONE = 'none';
    return Config;
})();
module.exports = Config;
