// asyncawait/comparison/mockfs.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

﻿var mock = require('mock-fs');
var Promise = require('bluebird');


var config = {
    'mocks/m1': {
        dir11: {
            file111: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
            file112: 'bbbbbbbbbbbbbb',
            file113: 'ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc',
            file114: 'dddddddddddddddddddddddddddddddddddddddd',
            file115: 'eeeeeee',
            file116: 'ffffffffffffffffff',
            file117: 'gggggggggggggggggggggggggggggggggggggggggggggggggggg',
            file118: 'hhhhhhhhhhhhhhhhhhhhhhhhhhh',
            file119: 'iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii',
        },
        dir12: {},
        file11: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
        file12: 'b',
        file13: 'ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc',
    },
    'mocks/m2': {
        file21: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
    },
    'mocks/m3': {
    }
};


var fs = Promise.promisifyAll(mock.fs(config));
module.exports = fs;
