// cliclopts/index.js
// Copyright (c) 2017 The Authors of 'JWTS for NODE'
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

function Cliclopts (opts) {
  if (!(this instanceof Cliclopts)) {
    return new Cliclopts(opts)
  }
  this.opts = opts || []
  this.indent = '    '
  this.width = 23
}

Cliclopts.prototype.print = function (stream) {
  var output = stream || process.stdout
  output.write(this.usage())
}

Cliclopts.prototype.usage = function () {
  var output = ''

  this.opts
    .filter(function (opt) {
      return opt.help
    })
    .forEach(function (option) {
      output += this.indent

      var helpIndex = option.helpIndex

      if (!helpIndex) {
        helpIndex = '--' + option.name
        if (option.abbr) helpIndex += ', -' + option.abbr
      }
      output += helpIndex

      var offset = 0
      if (helpIndex.length > this.width) {
        output += '\n' + this.indent
      } else {
        offset = helpIndex.length
      }

      output += Array(this.width - offset).join(' ')

      output += option.help
      if (option.hasOwnProperty('default')) {
        output += ' (default: ' + JSON.stringify(option.default) + ')'
      }
      output += '\n'
    }.bind(this))

  return output
}

Cliclopts.prototype.booleans = function () {
  return this.opts
    .filter(function (opt) {
      return opt.boolean
    })
    .map(function (opt) {
      return opt.name
    })
}

Cliclopts.prototype.boolean = Cliclopts.prototype.booleans

Cliclopts.prototype.default = function () {
  var defaults = {}
  this.opts.forEach(function (opt) {
    if ('default' in opt) {
      defaults[opt.name] = opt.default
      if ('abbr' in opt) defaults[opt.abbr] = opt.default
    }
  })
  return defaults
}

Cliclopts.prototype.alias = function () {
  var alias = {}
  this.opts.forEach(function (opt) {
    var build = []
    if ('alias' in opt) {
      if (typeof opt.alias === 'string') build.push(opt.alias)
      else build = build.concat(opt.alias)
    }
    if ('abbr' in opt) build.push(opt.abbr)
    alias[opt.name] = build
  })
  return alias
}

Cliclopts.prototype.options = function () {
  return {
    alias: this.alias(),
    boolean: this.boolean(),
    default: this.default()
  }
}

module.exports = Cliclopts
